server/server.go

```diff
@@ -352,10 +332,13 @@ func (s *Server) Run() error {
 			}
 
 			logutil.BgLogger().Error("accept failed", zap.Error(err))
-			return errors.Trace(err)
```

Comment:
I think we still need to return errors, in `tidb-server`, there is a non-nil assertion for `server.Run`, which will raise a negative return code to the terminal. The current modification actually changes this behavior.

In case of errors, and user did not press `ctrl+c`, the program will hang due to `<-exited` channel:

https://github.com/pingcap/tidb/blob/master/tidb-server/main.go#L193-L200

Reply:
So I basically need to have a channel for returning any errors and probably a WaitGroup to know when both are done? So if error, just return it otherwise wait until the the shutdown process has closed the listening sockets which will stop the `startNetworkListener`.

Reply:
I just used an error channel and waited for either an error or both to finish. Please take a look if it is what you expected?

Reply:
Yeah, it works now.